syntax = "proto3";

package statistics;

service StatisticsService {
  rpc GetTaskStats (TaskRequest) returns (TaskStatsResponse);
  rpc GetTopTasks (TopTasksRequest) returns (TopTasksResponse);
  rpc GetTopUsers (TopUsersRequest) returns (TopUsersResponse);
}

message TaskRequest {
  string task_id = 1;
}

message TaskStatsResponse {
  string task_id = 1;
  int32 views = 2;
  int32 likes = 3;
}

message TopTasksRequest {
  // "views" or "likes"
  string sort_by = 1;
}

message TopTasksResponse {
  repeated Task task = 1;
}

message Task {
  string task_id = 1;
  string author = 2;
  int32 count = 3;
}

message TopUsersRequest {}

message TopUsersResponse {
  repeated User user = 1;
}

message User {
  string login = 1;
  int32 total_likes = 2;
}
